Simplified/MyBooks/NYPLMyBooksDownloadCenter.m

```diff
@@ -607,66 +606,116 @@ - (void)returnBookWithIdentifier:(NSString *)identifier
     [[NYPLADEPT sharedInstance] returnLoan:fulfillmentId
                                     userID:[[NYPLUserAccount sharedAccount] userID]
                                   deviceID:[[NYPLUserAccount sharedAccount] deviceID]
-                                completion:^(BOOL success, __unused NSError *error) {
-                                  if(!success) {
+                                completion:^(BOOL success, NSError *error) {
+                                  if(success) {
+                                    [self revokeLoanAndRemoveBook:book state:state];
+                                  } else {
                                     NYPLLOG(@"Failed to return loan via NYPLAdept.");
+                                    [self presentAlertForError:error
+                                                   orErrorDict:nil
+                                                 returningBook:book];
                                   }
                                 }];
+  } else {
+    // if we have no fulfillment ID or we don't require authentication, we just
+    // need to let the Circulation Manager server know and remove the book
+    [self revokeLoanAndRemoveBook:book state:state];
   }
+#else
+  [self revokeLoanAndRemoveBook:book state:state];
 #endif
+}
 
+- (void)revokeLoanAndRemoveBook:(NYPLBook *)book state:(NYPLBookState)state
+{
+  BOOL didDownloadBook = (state == NYPLBookStateDownloadSuccessful
+                          || state == NYPLBookStateUsed);
+  NSString *const identifier = book.identifier;
+
+  // maybe this is the case when book has no DRM so there's no loan to revoke?
   if (!book.revokeURL) {
-    if (downloaded) {
+    if (didDownloadBook) {
       [self deleteLocalContentForBookIdentifier:identifier];
     }
     [[NYPLBookRegistry sharedRegistry] removeBookForIdentifier:identifier];
     [[NYPLBookRegistry sharedRegistry] save];
-  } else {
-    [[NYPLBookRegistry sharedRegistry] setProcessing:YES forIdentifier:book.identifier];
-    [NYPLOPDSFeed withURL:book.revokeURL shouldResetCache:NO completionHandler:^(NYPLOPDSFeed *feed, NSDictionary *error) {
+    return;
+  }
 
-      [[NYPLBookRegistry sharedRegistry] setProcessing:NO forIdentifier:book.identifier];
-      
-      if(feed && feed.entries.count == 1)  {
-        NYPLOPDSEntry *const entry = feed.entries[0];
-        if(downloaded) {
-          [self deleteLocalContentForBookIdentifier:identifier];
-        }
-        NYPLBook *returnedBook = [NYPLBook bookWithEntry:entry];
-        if(returnedBook) {
```

Comment:
I removed this `if` because it's irrelevant. It's impossible to get a nil book from `bookWithEntry` if `entry` is not nil, and we know that from `NYPLOPDSEntry *const entry = feed.entries[0];` (NSArrays cannot contain nil values).